src/preamble.js

```diff
@@ -1669,6 +1669,46 @@ function lookupSymbol(ptr) { // for a pointer, print out all symbols that resolv
 var memoryInitializer = null;
 
 #if USE_PTHREADS
+
+// JS library / EM_ASM proxying
+
+function proxyToMainThread() {
+  var args = Array.prototype.slice.call(arguments);
+  // Arguments are:
+  //   * the index of the function
+  //   * the call args
+  //   * whether it is sync
+  var sync = args[args.length - 1];
+  // The serialization buffer contains the number of call params, and then
+  // all the args here.
+  // We also pass 'sync' to C separately, since C needs to look at it.
+  // The buffer remains alive until receiveOnMainThread frees it.
+  var bufferLen = args.length + 1;
+  var buffer = _malloc(bufferLen * 8); // TODO: stackAlloc if sync?
+  var numCallArgs = args.length - 2;
+  HEAPF64[buffer >> 3] = numCallArgs; // num of call args
+  for (var i = 0; i < bufferLen - 1; i++) {
+    HEAPF64[(buffer >> 3) + 1 + i] = args[i];
+  }
+  return _emscripten_run_in_main_runtime_thread_js(buffer, sync);
+}
+
+function receiveOnMainThread(buffer) {
+  var numCallArgs = HEAPF64[buffer >> 3];
+  var index = HEAPF64[(buffer >> 3) + 1];
+  var callArgs = HEAPF64.subarray((buffer >> 3) + 2, (buffer >> 3) + 2 + numCallArgs);
+  var sync = HEAPF64[(buffer >> 3) + 2 + numCallArgs];
+  // Proxied JS library funcs are encoded as positive values, and
+  // EM_ASMs as negative values (see include_asm_consts)
+  if (index > 0) {
+    ret = proxiedFunctionTable[index].apply(null, callArgs);
+  } else {
+    ret = ASM_CONSTS[-index - 1].apply(null, callArgs);
```

Comment:
The `HEAPF64.subarray() and `func.apply(null, callArgs)` constructs make me feel uneasy, this occurs very often for proxied WebGL functions. I think we should look into retaining the old form of `receiveMainThread()` function with the tight switch-case for each argument count as an option for garbageless execution and comparison.

Reply:
If we are really worried about the garbage, we can reuse a single JS array here and resize it each time. This still avoids a switch-case for each argument count (and works for an arbitrary number of arguments). Adding a commit with that.

Reply:
Not sure that kind of pattern would feel much different, there are no guarantees how resizing an array works with respect to allocation internally, the `.apply()` still pattern still exists, and then on top there will have to be a for loop to copy each arg over.

Reply:
No guarantees, but array resizing has been heavily optimized for many years in JS engines - it's an ancient part of the language. `.apply()` has no significant cost, not sure what the concern there is? It's true there's a loop, but I think the overhead there is negligible compared to the cost of proxying.

I can benchmark this if you want, but this type of code is very common in JS, and VMs optimize it well. This kind of thing should not be a perf problem (if it is, it should be easy to measure and motivate a change).

The reason I'm pushing back here is that the switch pattern is significantly less readable. Less code is better and safer I feel.